oneargopy/Argo.py

```diff
@@ -0,0 +1,279 @@
+                    directory_path.mkdir()
+                except Exception as e:
+                    if self.download_settings.verbose: print(f'Failed to create the {directory} directory: {e}')
+
+
+    def __download_index_file(self, file_name: str) -> None:
+        """ A function to download and save an index file from GDAC sources. 
+
+            :param: filename : str - The name of the file we are downloading.
+        """
+        index_directory = Path(self.download_settings.base_dir.joinpath("Index"))
+
+        # Get the expected filepath for the file
+        file_path = index_directory.joinpath(file_name)
+
+        # Check if the filepath exists
+        if file_path.exists():
+
+            # Check if the settings allow  for updates
+            if self.download_settings.update == 0:
+                if self.download_settings.verbose: 
+                    print(f'The download settings have update set to 0, indicating that we do not want to update index files.')
+            else: 
+                txt_last_modified_time = Path(file_path).stat().st_mtime
+                current_time = datetime.now().timestamp()
+                txt_seconds_since_modified = current_time - txt_last_modified_time
+
+                if self.download_settings.verbose: print(f'It has been {txt_seconds_since_modified}s since {file_name} was modified.')
+                if self.download_settings.verbose: print(f'The download threshold is {self.download_settings.update}')
+
+                # Check if the file should be updated
+                if (txt_seconds_since_modified > self.download_settings.update):
+                    if self.download_settings.verbose: print(f'Updating {file_name}...')
+                    self.__try_download(file_name ,True)
+                else:
+                    if self.download_settings.verbose: print(f'{file_name} does not need to be updated yet.\n')
+
+        # if the file doesn't exist then download it
+        else: 
+            if self.download_settings.verbose: print(f'{file_name} needs to be downloaded.')
+            self.__try_download(file_name, False)
+
+
+    def __try_download(self, file_name: str, update_status: bool)-> None:
+        """ A function that attempts to download a file from both GDAC sources.
+
+            :param: file_name : str - The name of the file to download
+            :param: update_status: bool - True if the file exists and we 
+                are trying to update it. False if the file hasn't been 
+                downloaded yet. 
+        """
+        index_directory = Path(self.download_settings.base_dir.joinpath("Index"))
+
+        success = False
+        iterations = 0
+        txt_save_path = index_directory.joinpath(file_name)
+        gz_save_path = index_directory.joinpath("".join([file_name, ".gz"]))
+
+        while (not success) and (iterations < self.download_settings.max_attempts):
+            # Try both hosts (perfered one is listed first in download settings)
```

Comment:
preferred

Reply:
![image](https://github.com/lovelysandlonelys/OneArgo-Python-Dev/assets/91633941/259909a9-671b-4944-a557-9d1ee5af59c0)

Reply:
![image](https://github.com/lovelysandlonelys/OneArgo-Python-Dev/assets/91633941/7e28cc94-a071-42ff-b47a-620aac899165)
fixed here and in functions